Refactor health structure and configuration

RVanhuysseXenit opened the pull request:
Fixes https://github.com/xenit-eu/alfresco-actuators/issues/6

Changes include:
* restructuring the health services for more granular checks
* refactoring endpoint to provided details to authenticated administrators
* add configuration for the interpretation of the readonly status in relation to OVAMSLA-111

tgeens commented on alfresco-actuators/src/main/java/eu/xenit/actuators/services/StatusInfoService.java:
```diff
@@ -0,0 +1,83 @@
+            boolean healthyOnReadonly = Boolean.parseBoolean(
+                    (String) globalProperties.getOrDefault(GLOBAL_HEALTH_ON_READONLY, "true"));
+            if (!healthyOnReadonly && statusInfo.getReadOnly()) {
+                throw new HealthException("Alfresco is in readonly. Handling as down.");
```
Don't use exceptions for control flow:
* move the statements of the catch-block in here
* remove `HealthException` class again, it should not exist

tgeens commented on alfresco-actuators/src/main/java/eu/xenit/actuators/services/StatusInfoService.java:
```diff
@@ -0,0 +1,83 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(StatusInfoService.class);
+
+    public static final String GLOBAL_HEALTH_ON_READONLY = "eu.xenit.actuators.repository.healthyOnReadOnly";
```
I'm not sure I agree with using a config flag here. The reason you have this flag, is that you don't want your service getting kicked out of the load balancer, because it's going in read-only ?

Design goal: Organize health indicators into groups that can be used for different purposes. The load balancer can track one healthgroup, the orchestrator a different one.

(Health-group support should probably be a different PR, but still - if you include that flag now, someone is going to get surprised and upset, if it's removed in the future)

References:
* https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.health.groups
* https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.spring-application.application-availability

RVanhuysseXenit commented:
* Added a grouping system that loads a groupdefinition at startup.
* Removed check & flag for readonly. Will define separate HealthIndicator + group definition in separate project.
* defined default groups for readiness & liveliness

Todos (for other pr)
* dynamically define or load groups
* merge group configs at startup

Predicted sideeffects:
The basic health endpoint iterates over all healthIndicators. Conflicting indicators may result in undesired global status.

RVanhuysseXenit commented:
Closing pr:
* moving groups config to different branch/pr
* including changes required to add in custom services (visibility, splitup of exisiting classes) in new pr
